includes/class-wc-stripe-payment-tokens.php

```diff
@@ -262,28 +267,33 @@ public function woocommerce_get_customer_upe_payment_tokens( $tokens, $user_id,
 			}
 		}
 
-		foreach ( $retrievable_payment_method_types as $payment_method_id ) {
-			$payment_methods = $customer->get_payment_methods( $payment_method_id );
-
-			// Prevent unnecessary recursion, WC_Payment_Token::save() ends up calling 'woocommerce_get_customer_payment_tokens' in some cases.
-			remove_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
-			foreach ( $payment_methods as $payment_method ) {
-				if ( ! isset( $remaining_tokens[ $payment_method->id ] ) ) {
-					$payment_method_type = $this->get_original_payment_method_type( $payment_method );
-					if ( ! in_array( $payment_method_type, $reusable_payment_methods, true ) ) {
-						continue;
+		try {
+			foreach ( $retrievable_payment_method_types as $payment_method_id ) {
+				$payment_methods = $customer->get_payment_methods( $payment_method_id );
+
+				// Prevent unnecessary recursion, WC_Payment_Token::save() ends up calling 'woocommerce_get_customer_payment_tokens' in some cases.
+				remove_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
+				foreach ( $payment_methods as $payment_method ) {
+					if ( ! isset( $remaining_tokens[ $payment_method->id ] ) ) {
+						$payment_method_type = $this->get_original_payment_method_type( $payment_method );
+						if ( ! in_array( $payment_method_type, $reusable_payment_methods, true ) ) {
+							continue;
+						}
+						// Create new token for new payment method and add to list.
+						$upe_payment_method         = $gateway->payment_methods[ $payment_method_type ];
+						$token                      = $upe_payment_method->create_payment_token_for_user( $user_id, $payment_method );
+						$tokens[ $token->get_id() ] = $token;
+					} else {
+						// Count that existing token for payment method is still present on Stripe.
+						// Remaining IDs in $remaining_tokens no longer exist with Stripe and will be eliminated.
+						unset( $remaining_tokens[ $payment_method->id ] );
 					}
-					// Create new token for new payment method and add to list.
-					$upe_payment_method         = $gateway->payment_methods[ $payment_method_type ];
-					$token                      = $upe_payment_method->create_payment_token_for_user( $user_id, $payment_method );
-					$tokens[ $token->get_id() ] = $token;
-				} else {
-					// Count that existing token for payment method is still present on Stripe.
-					// Remaining IDs in $remaining_tokens no longer exist with Stripe and will be eliminated.
-					unset( $remaining_tokens[ $payment_method->id ] );
 				}
+				add_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
 			}
-			add_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
+		} catch ( WC_Stripe_Exception $e ) {
+			wc_add_notice( $e->getLocalizedMessage(), 'error' );
```

Comment:
Just curious. Why do we add a WC notice in this and woocommerce_get_customer_payment_tokens_legacy cases, but not in other of the exceptions we're catching?

Reply:
These 2 will add the error notice on the checkout page. I have noticed that if anything goes wrong, we show an error notice on the checkout page. For this reason, I have added these 2 notices here. :sweat_smile: